File: src/c/autonomous_trust/replication/bisection.c
/********************
 *  Copyright 2026 TekFive, Inc., Sean M. Brennan, and contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 *******************/

#include "replication/bisection.h"

#include <string.h>

#include <sodium.h>

/* blake2b(in) -> 64 lowercase hex chars + NUL, the same primitive as the
 * reputation Merkle's merkle_hash_hex (crypto_generichash 32-byte digest,
 * sodium_bin2hex), i.e. Python MerkleTree.get_hash. */
static void _hash_hex(const unsigned char *in, size_t len,
                      char out[AT_REPL_HASH_HEX_LEN + 1])
{
    unsigned char digest[32];
    crypto_generichash(digest, sizeof(digest), in, len, NULL, 0);
    sodium_bin2hex(out, AT_REPL_HASH_HEX_LEN + 1, digest, sizeof(digest));
}

/* h_0 = H(s_0); h_i = H(h_{i-1} || s_i). Fills chain[0..n-1]. The previous
 * digest is prepended as its 64 hex bytes, exactly as the Python twin chains
 * the hex-ASCII digest with the next state's UTF-8 bytes. */
static void _hash_chain(const char *const *states, size_t n,
                        char (*chain)[AT_REPL_HASH_HEX_LEN + 1])
{
    unsigned char buf[AT_REPL_HASH_HEX_LEN + 512];
    for (size_t i = 0; i < n; i++)
    {
        const char *s = states[i] ? states[i] : "";
        size_t slen = strlen(s);
        size_t prev_len = (i == 0) ? 0 : AT_REPL_HASH_HEX_LEN;
        if (prev_len + slen > sizeof(buf))
            slen = sizeof(buf) - prev_len;   /* defensive; tokens are short */
        if (i != 0)
            memcpy(buf, chain[i - 1], prev_len);
        memcpy(buf + prev_len, s, slen);
        _hash_hex(buf, prev_len + slen, chain[i]);
    }
}

bool at_replication_commit_root(const char *const *states, size_t n,
                                char out[AT_REPL_HASH_HEX_LEN + 1])
{
    if (n == 0 || states == NULL || n > AT_REPL_MAX_STEPS)
        return false;
    static char chain[AT_REPL_MAX_STEPS][AT_REPL_HASH_HEX_LEN + 1];
    _hash_chain(states, n, chain);
    memcpy(out, chain[n - 1], AT_REPL_HASH_HEX_LEN + 1);
    return true;
}

/* Smallest index where two chains differ, by binary search over the monotonic
 * "differ" predicate. Returns the index and sets *found; when the chains agree
 * over their common length, *found is true only if the lengths differ (one
 * kept computing past the other), and the index is that common length. */
static int _first_divergence(const char (*chain_a)[AT_REPL_HASH_HEX_LEN + 1],
                             size_t na,
                             const char (*chain_b)[AT_REPL_HASH_HEX_LEN + 1],
                             size_t nb, bool *found)
{
    size_t n = na < nb ? na : nb;
    if (n == 0 || strcmp(chain_a[n - 1], chain_b[n - 1]) == 0)
    {
        if (na != nb)
        {
            *found = true;
            return (int)n;
        }
        *found = false;
        return 0;
    }
    size_t lo = 0, hi = n - 1;
    while (lo < hi)
    {
        size_t mid = (lo + hi) / 2;
        if (strcmp(chain_a[mid], chain_b[mid]) != 0)
            hi = mid;
        else
            lo = mid + 1;
    }
    *found = true;
    return (int)lo;
}

/* Python None == None is true, None == x is false: a state past the end of a
 * trace is NULL and matches only another NULL. */
static bool _streq_nullable(const char *a, const char *b)
{
    if (a == NULL || b == NULL)
        return a == b;
    return strcmp(a, b) == 0;
}

bool at_replication_bisect_adjudicate(
    const char *const *states_a, size_t na,
    const char *const *states_b, size_t nb,
    const char *const *reference, size_t nref,
    int *divergence_out, bool *diverged_out,
    at_repl_verdict_t *verdict_a, at_repl_verdict_t *verdict_b)
{
    if (na > AT_REPL_MAX_STEPS || nb > AT_REPL_MAX_STEPS
        || nref > AT_REPL_MAX_STEPS)
        return false;
    static char chain_a[AT_REPL_MAX_STEPS][AT_REPL_HASH_HEX_LEN + 1];
    static char chain_b[AT_REPL_MAX_STEPS][AT_REPL_HASH_HEX_LEN + 1];
    _hash_chain(states_a, na, chain_a);
    _hash_chain(states_b, nb, chain_b);

    bool found = false;
    int k = _first_divergence(chain_a, na, chain_b, nb, &found);
    if (!found)
    {
        if (diverged_out) *diverged_out = false;
        if (divergence_out) *divergence_out = -1;
        if (verdict_a) *verdict_a = AT_REPL_CORROBORATED;
        if (verdict_b) *verdict_b = AT_REPL_CORROBORATED;
        return true;
    }

    const char *ref_k = ((size_t)k < nref) ? reference[k] : NULL;
    const char *a_k = ((size_t)k < na) ? states_a[k] : NULL;
    const char *b_k = ((size_t)k < nb) ? states_b[k] : NULL;
    if (diverged_out) *diverged_out = true;
    if (divergence_out) *divergence_out = k;
    if (verdict_a)
        *verdict_a = _streq_nullable(a_k, ref_k) ? AT_REPL_CORROBORATED
                                                 : AT_REPL_OUTVOTED;
    if (verdict_b)
        *verdict_b = _streq_nullable(b_k, ref_k) ? AT_REPL_CORROBORATED
                                                 : AT_REPL_OUTVOTED;
    return true;
}

File: src/autonomous-trust/autonomous_trust/core/_python/replication/bisection.py
# ******************
#  Copyright 2026 TekFive, Inc., Sean M. Brennan, and contributors
#
#   Licensed under the Apache License, Version 2.0 (the "License");
#   you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#   See the License for the specific language governing permissions and
#   limitations under the License.
# ******************
"""The bisection dispute game for a replicated task (R+D.md §12.6, slice C).

When two executors disagree about a long computation, :mod:`.adjudication`
returns ``dispute`` and scores nobody: a majority is not an oracle, and one
replica's word must not defame the executor it checked. This resolves that
dispute the way Truebit and optimistic-rollup fraud proofs do -- by localizing
the fault to a single step rather than re-running the whole computation, at a
cost logarithmic in its length.

**Why a hash chain, not a list of states.** Each executor commits to a chained
digest ``h_i = H(h_{i-1} || s_i)`` (``h_0 = H(s_0)``, the shared input). The
chaining is the whole point: because ``h_i`` depends on every state up to ``i``,
"the two chains differ at index ``i``" is MONOTONIC -- once they differ they
differ forever -- which is exactly what lets a binary search find the FIRST
divergent step in ``O(log n)`` challenges without either executor revealing more
than one digest per round. A bare list of states gives no such monotonicity: two
states can differ at ``k`` and agree again at ``k+1``, and no binary search is
sound over it. The primitive is the codebase's Merkle hash --- blake2b, 64-char
lowercase hex --- so a committed root is identical in the C twin
(``src/c/autonomous_trust/replication/bisection.c``).

**What the conformance corpus pins, and what it does not.** It pins the game:
the committed root, the divergence index the search finds, and the verdict the
single-step check renders against a reference trace that stands in for
"re-execute step k". It does NOT wire this to live tasks, because the game
requires DETERMINISTIC REPLAY of a peer's work -- a real constraint on how a
capability is written (the doc is explicit), and one AT cannot impose on an
arbitrary capability unilaterally. The live path is gated on that precondition;
see doc/architecture/replication.md.

**States are opaque tokens.** An intermediate state is compared for exact
equality, and hashed as its UTF-8 bytes. Numeric tolerance belongs to a task's
final RESULT (slice B), not to the bit-exact intermediate states a
deterministic replay produces; a step either reproduces or it does not.

The verdicts are :mod:`.adjudication`'s own -- ``corroborated`` / ``outvoted``
-- so :func:`.adjudication.verify` maps a resolved dispute to the same
``replication``-channel scores an outright majority would.
"""

from __future__ import annotations

from typing import Any, Optional

from autonomous_trust.core.structures.merkle import MerkleTree

from .adjudication import CORROBORATED, OUTVOTED


def _state_bytes(state: Any) -> bytes:
    """A state token's canonical bytes for hashing: its UTF-8 text."""
    if isinstance(state, bytes):
        return state
    return str(state).encode('utf-8')


def hash_chain(states: list[Any]) -> list[str]:
    """The committed chain ``h_0 .. h_{n-1}`` as 64-char lowercase-hex strings.

    ``h_0 = H(s_0)`` and ``h_i = H(h_{i-1} || s_i)``, the same
    canonical-content-chained-with-previous form the reputation Merkle entry
    uses, so the digest is identical in both runtimes.
    """
    chain: list[str] = []
    prev = b''
    for state in states:
        digest = MerkleTree.get_hash(prev + _state_bytes(state))  # hex bytes
        chain.append(digest.decode('ascii'))
        prev = digest
    return chain


def commit_root(states: list[Any]) -> Optional[str]:
    """The executor's committed root: the final chain digest, or ``None`` for
    an empty trace (nothing was computed to commit to)."""
    chain = hash_chain(states)
    return chain[-1] if chain else None


def first_divergence(chain_a: list[str], chain_b: list[str]) -> Optional[int]:
    """The first index at which two committed chains differ, by binary search.

    ``None`` when neither chain diverges from the other over their common
    length AND they are the same length; when they agree over the common length
    but one is longer, the divergence is at that common length (one executor
    kept computing past where the other stopped). Relies on the chain's
    monotonic "differ" predicate; a raw state list could not be searched this
    way.
    """
    n = min(len(chain_a), len(chain_b))
    if n == 0 or chain_a[n - 1] == chain_b[n - 1]:
        return n if len(chain_a) != len(chain_b) else None
    # differ(n-1) is true and differ is monotonic: find the smallest true.
    lo, hi = 0, n - 1
    while lo < hi:
        mid = (lo + hi) // 2
        if chain_a[mid] != chain_b[mid]:
            hi = mid
        else:
            lo = mid + 1
    return lo


def bisect_adjudicate(peer_a: str, states_a: list[Any],
                      peer_b: str, states_b: list[Any],
                      reference: list[Any]
                      ) -> tuple[Optional[int], dict[str, str]]:
    """Resolve a two-way dispute by localizing and checking the first divergent
    step.

    Returns ``(divergence_index, {peer: verdict})``. Both executors agree at the
    input, so the search finds the first step where their committed chains part;
    the reference's state at that step is the ground truth a re-execution would
    produce, and whichever executor reproduced it is ``corroborated`` while one
    that did not is ``outvoted``. Both can be outvoted -- two executors wrong in
    different ways at the same step -- which is the honest outcome, not a tie.

    ``divergence_index`` is ``None`` only when the two never diverge, in which
    case there was no dispute to resolve and both are ``corroborated``.
    """
    chain_a = hash_chain(states_a)
    chain_b = hash_chain(states_b)
    k = first_divergence(chain_a, chain_b)
    if k is None:
        return None, {peer_a: CORROBORATED, peer_b: CORROBORATED}

    ref_k = reference[k] if k < len(reference) else None
    a_k = states_a[k] if k < len(states_a) else None
    b_k = states_b[k] if k < len(states_b) else None
    return k, {
        peer_a: CORROBORATED if a_k == ref_k else OUTVOTED,
        peer_b: CORROBORATED if b_k == ref_k else OUTVOTED,
    }
